torch/csrc/api/include/torch/data/dataloader.h

```diff
@@ -49,6 +50,37 @@ class DataLoader {
     }
   }
 
+  /// Constructs a new `DataLoader` from a `dataset`, `options`, and `Sampler`.
```

Comment:
So, for this whole thing, if you're concerned about copying the dataset, can't you just do:

```
struct MyVeryExpensiveToCopyDataset : ... { };
struct MyCheapToCopyDataset {
std::shared_ptr<MyVeryExpensiveToCopyDataset> dataset;
};

auto dataloader = 
    torch::data::make_data_loader(
        MyCheapToCopyDataset{std::make_shared<MyVeryExpensiveToCopyDataset>()});
```

and then we could just re-use the previous constructor and not need any custom code at all?

Reply:
Then we will have to modify the existing constructor to create the main_thread_dataset conditionally when the number of workers are more than 0, also if we need to validate the arguments now they will be conditional at the constructor with chunk support or not, I think sooner or later it will be ugly. That is the reason for the new constructor with a clear flag  indicating the chunk support.

Reply:
But in that case we don't need a dataset in the main thread anymore I think? You'd just have the workers reference the wrapper over the expensive dataset, and behind the scenes they would be accessing the shared dataset just like in your proposal. It would also work very nicely with transforms, since you would just apply the transforms on top of the wrapper dataset, such that they would be happening in the worker threads and not inside the shared dataset. What do you think? We could code up how that would look like, but I think it sounds doable

Reply:
I see your point, but unless we ask this from the user, we will end up copying the dataset one more time when creating the shared_ptr and assuming a proper copy constructor is defined in the dataset (I could not find a torch version of make_shared with move capability). In any case, do you think this complexity is needed for this change. We can get back to this, if it simplify things when we have a complete implementation. Currently this is only a part of it. What do you think?

btw; even with the current PR, the map() operations are happening in the dataloader's threads. Only the prefetching is happening in the dataset.